snuba/request/validation.py

```diff
@@ -116,64 +118,24 @@ def build_request(
     with sentry_sdk.start_span(description="build_request", op="validate") as span:
         try:
             request_parts = schema.validate(body)
-            tenant_referrer = request_parts.attribution_info["tenant_ids"].get(
-                "referrer"
-            )
-            if tenant_referrer != referrer:
-                metrics.increment(
-                    "referrer_mismatch",
-                    tags={
-                        "tenant_referrer": tenant_referrer or "none",
-                        "request_referrer": referrer,
-                    },
+            referrer = _get_referrer(request_parts, referrer)
+            settings_obj = _get_settings_object(settings_class, request_parts, referrer)
+            try:
+                query, snql_anonymized = parser(
+                    request_parts, settings_obj, dataset, custom_processing
                 )
-            # Handle an edge case where the legacy endpoint is used.
-            referrer = tenant_referrer or referrer
-
-            if settings_class == HTTPQuerySettings:
-                query_settings: MutableMapping[str, bool | str] = {
-                    **request_parts.query_settings,
-                    "consistent": _consistent_override(
-                        request_parts.query_settings.get("consistent", False), referrer
-                    ),
-                }
-                query_settings["referrer"] = referrer
-                # TODO: referrer probably doesn't need to be passed in, it should be from the body
-                settings_obj: Union[HTTPQuerySettings, SubscriptionQuerySettings]
-                # the parameters accept either `str` or `bool` but we pass in `str | bool`
-                settings_obj = settings_class(**query_settings)  # type: ignore
-            elif settings_class == SubscriptionQuerySettings:
-                settings_obj = settings_class(
-                    consistent=_consistent_override(True, referrer),
+            except PostProcessingError as exception:
```

Comment:
Could we add a test to make sure the recursion query is being handled properly and that we're catching `PostProcessingError`?

Reply:
done